common/src/main/java/com/microsoft/identity/common/internal/controllers/ExceptionAdapter.java

```diff
@@ -137,111 +130,131 @@ public static BaseException exceptionFromAcquireTokenResult(final AcquireTokenRe
                 );
             }
 
-            if (null == outErr) {
-                outErr = new ServiceException(
+            outErr = getExceptionFromOAuthError(
                         tokenErrorResponse.getError(),
+                        tokenErrorResponse.getSubError(),
                         tokenErrorResponse.getErrorDescription(),
-                        null
-                );
-            }
+                    synthesizeHttpResponse(tokenErrorResponse.getStatusCode(),
+                            tokenErrorResponse.getResponseHeadersJson(),
+                            tokenErrorResponse.getResponseBody()));
 
-            applyCliTelemInfo(tokenResult, outErr);
+            applyCliTelemInfo(tokenResult.getCliTelemInfo(), outErr);
+        }
 
-            applyHttpErrorResponseData(
-                    outErr,
-                    tokenErrorResponse.getStatusCode(),
-                    tokenErrorResponse.getResponseHeadersJson(),
-                    tokenErrorResponse.getResponseBody()
-            );
+        return null;
+    }
 
-            return outErr;
+    /**
+     * Determine if an exception owning the given error codes should be converted into UiRequiredException.
+     *
+     * @param oAuthError
+     * @param oAuthSubError
+     * @return boolean
+     * */
+    private static boolean shouldBeConvertedToUiRequiredException(final String oAuthError,
+                                                                  final String oAuthSubError){
+        // Invalid_grant doesn't necessarily requires UI protocol-wise.
+        // We simplify our logic because this layer is also used by MSAL.
+        if (oAuthError.equalsIgnoreCase(AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT)){
+            return true;
         }
 
-        return null;
+        return false;
+    }
+
+    /**
+     * Get an exception object from the given oAuth values.
+     *
+     * @param oAuthError
+     * @param oAuthSubError
+     * @param oAuthErrorDescription
+     * @param response
+     * @return ServiceException, UiRequiredException
+     * */
+    public static ServiceException getExceptionFromOAuthError(final String oAuthError,
```

Comment:
Is there a reason that this method doesn't just accept the TokenErrorResponse?  Since it's using all of the information from that object?

Reply:
This will also be called by PRTController (so that we have one place that does the exception conversion, which is ExceptionAdapter)